clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsGroupHeartbeatRequestManager.java

```diff
@@ -0,0 +1,387 @@
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Timer;
+
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+
+public class StreamsGroupHeartbeatRequestManager implements RequestManager {
+
+    static class HeartbeatState {
+
+        private final StreamsMembershipManager membershipManager;
+        private final int rebalanceTimeoutMs;
+
+        private final StreamsRebalanceData streamsRebalanceData;
+
+        public HeartbeatState(final StreamsRebalanceData streamsRebalanceData,
+                              final StreamsMembershipManager membershipManager,
+                              final int rebalanceTimeoutMs) {
+
+            this.membershipManager = membershipManager;
+            this.rebalanceTimeoutMs = rebalanceTimeoutMs;
+            this.streamsRebalanceData = streamsRebalanceData;
+        }
+
+        public StreamsGroupHeartbeatRequestData buildRequestData() {
+            StreamsGroupHeartbeatRequestData data = new StreamsGroupHeartbeatRequestData();
+            data.setGroupId(membershipManager.groupId());
+            data.setMemberId(membershipManager.memberId());
+            data.setMemberEpoch(membershipManager.memberEpoch());
+            membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+            StreamsGroupHeartbeatRequestData.Topology topology = new StreamsGroupHeartbeatRequestData.Topology();
+            topology.setSubtopologies(getTopologyFromStreams());
+            topology.setEpoch(streamsRebalanceData.topologyEpoch());
+            data.setRebalanceTimeoutMs(rebalanceTimeoutMs);
+            data.setTopology(topology);
+            data.setProcessId(streamsRebalanceData.processId().toString());
+            data.setActiveTasks(Collections.emptyList());
+            data.setStandbyTasks(Collections.emptyList());
+            data.setWarmupTasks(Collections.emptyList());
+            streamsRebalanceData.endpoint().ifPresent(userEndpoint -> {
+                data.setUserEndpoint(new StreamsGroupHeartbeatRequestData.Endpoint()
+                    .setHost(userEndpoint.host())
+                    .setPort(userEndpoint.port())
+                );
+            });
+            data.setClientTags(streamsRebalanceData.clientTags().entrySet().stream()
+                .map(entry -> new StreamsGroupHeartbeatRequestData.KeyValue()
+                    .setKey(entry.getKey())
+                    .setValue(entry.getValue())
+                )
+                .collect(Collectors.toList()));
+            if (streamsRebalanceData.shutdownRequested()) {
+                data.setShutdownApplication(true);
```

Comment:
Could we simplify to simply `data.setShutdownApplication(streamsRebalanceData.shutdownRequested())` ?

Reply:
Good idea!